Source/Csla.AspNetCore/Blazor/ApplicationContextManagerBlazor.cs

```diff
@@ -0,0 +1,166 @@
+﻿#if NET5_0_OR_GREATER
+//-----------------------------------------------------------------------
+// <copyright file="ApplicationContextManager.cs" company="Marimer LLC">
+//     Copyright (c) Marimer LLC. All rights reserved.
+//     Website: https://cslanet.com
+// </copyright>
+// <summary>Application context manager that uses HttpContextAccessor</summary>
+//-----------------------------------------------------------------------
+using Csla.Core;
+using Microsoft.AspNetCore.Components.Authorization;
+using System;
+using System.Security.Claims;
+using System.Security.Principal;
+using System.Threading.Tasks;
+
+namespace Csla.AspNetCore.Blazor
+{
+  /// <summary>
+  /// Application context manager that uses HttpContextAccessor when 
+  /// resolving HttpContext to store context values.
+  /// </summary>
+  public class ApplicationContextManagerBlazor : IContextManager, IDisposable
+  {
+    private ContextDictionary LocalContext { get; set; }
+    private ContextDictionary ClientContext { get; set; }
+    private IPrincipal CurrentPrincipal { get; set; }
+    private readonly ClaimsPrincipal UnauthenticatedPrincipal = new();
+    private bool disposedValue;
+
+    /// <summary>
+    /// Gets the current HttpContext instance.
+    /// </summary>
+    protected AuthenticationStateProvider AuthenticationStateProvider { get; private set; }
+
+    /// <summary>
+    /// Gets or sets a reference to the current ApplicationContext.
+    /// </summary>
+    public ApplicationContext ApplicationContext { get; set; }
+
+    /// <summary>
+    /// Creates an instance of the object, initializing it
+    /// with the required IServiceProvider.
+    /// </summary>
+    /// <param name="authenticationStateProvider">AuthenticationStateProvider service</param>
+    public ApplicationContextManagerBlazor(AuthenticationStateProvider authenticationStateProvider)
+    {
+      AuthenticationStateProvider = authenticationStateProvider;
+      CurrentPrincipal = UnauthenticatedPrincipal;
+      AuthenticationStateProvider.AuthenticationStateChanged += AuthenticationStateProvider_AuthenticationStateChanged;
+      InitializeUser();
+    }
+
+    private void InitializeUser()
+    {
+      AuthenticationStateProvider_AuthenticationStateChanged(AuthenticationStateProvider.GetAuthenticationStateAsync());
```

Comment:
I'm still thinking this line will blow up with that GetState called before SetState error when the instigating code is started in a HostedService?

Reply:
I'm pretty sure you were seeing that because you were outside of Blazor. The point of these changes is to avoid the use of this code in that scenario.

Reply:
I will double check whether the error I saw only happens from HostedService calls...and I will report back here.
In the meantime, I did get around all the issues by registering my own `BlazorApplicationContextManager` (a copy of Rocky's) with the 3 small tweaks I mentioned in [this discussion](https://github.com/MarimerLLC/csla/discussions/2687#discussioncomment-2051923) and the small fix a couple comments later.

Reply:
Nope as I thought...the same unhandled error happens 
```cs        
     [2022-02-01T20:57:59.663Z] Error: System.InvalidOperationException: GetAuthenticationStateAsync was called before SetAuthenticationState.
 at Microsoft.AspNetCore.Components.Server.ServerAuthenticationStateProvider.GetAuthenticationStateAsync()
 at Csla.Blazor.ApplicationContextManager.InitializeUser()
```

simply from injecting the Csla.ApplicationContext into the Counter blazor page.
```cs
@inject ApplicationContext _appContext;
```

Project only references Csla.Blazor (not Csla.AspNet.Core).

Here is Program.cs
```cs
builder.Services.AddRazorPages();
builder.Services.AddServerSideBlazor();
builder.Services.AddHttpContextAccessor();

builder.Services.AddCsla(options =>
{
    options.AddServerSideBlazor();
});

```

Reply:
@rockfordlhotka and @TheCakeMonster - I've updated my [demo](https://github.com/swegele/Csla6BlazorSS) so it takes < 5 minutes to see this error.  

Totally removed HostedService.  Nothing funky.
Follow instructions on readme.md

1.  Run app, go to Counter page and see unhandled error (F12)
2. Stop app
3. Comment out injection line of my business object factory on Counter.razor
3. Run app again and see everything works

I'm not even calling the business objects...simply injecting one factory.  So weird.

Reply:
@swegele what is (in summary) the solution?

Reply:
It's a workaround for [this issue](https://github.com/dotnet/aspnetcore/issues/28684).  I openly admit I do not understand why it is happening in that very simple and clean demo.  If we can't figure that out, some order of operations thing I'm guessing, then we need a workaround (at least until Microsoft fixes that).

**Summary of workaround for a simple/default BSS contextmanager (not hybrid)** 

1. Make a version of an `ApplicationContextManager` for Blazor SS (separate from WASM) so we can count on the `AuthenticationStateProvider` implementing `IHostEnvironmentAuthenticationStateProvider`
2. Trap the error during `InitializeUser()` and call `SetAuthenticationState`.  That is what the error instructs us to do  
3. Also call the `SafeSetCurrentUser` method from SetUser

Credit to @TheCakeMonster as this is a slightly tweaked version of his code last week...so far it fixes everything in all scenarios for BSS that I've done so far.

```cs
private void InitializeUser()
{
        try
        {
            var getUserTask = AuthenticationStateProvider.GetAuthenticationStateAsync();
            AuthenticationStateProvider_AuthenticationStateChanged(getUserTask);
        }
        catch (InvalidOperationException ioex)
        {
            //Test for this specific error
            string message = ioex.Message;
            if (message.Contains(nameof(AuthenticationStateProvider.GetAuthenticationStateAsync))
                && message.Contains(nameof(IHostEnvironmentAuthenticationStateProvider.SetAuthenticationState)))
            {
                //error tells us no one has called set yet...so let's call it
                SafeSetCurrentUser(AuthenticationStateProvider, UnauthenticatedPrincipal);
            }
            else
            {
               thow;
            }
        }
    }
}

private void SafeSetCurrentUser(IHostEnvironmentAuthenticationStateProvider hostEnvironmentAuthProvider, ClaimsPrincipal principal)
{
    var task = new Task<AuthenticationState>(() => new AuthenticationState(principal));
    hostEnvironmentAuthProvider.SetAuthenticationState(Task.FromResult(authState));
}

public virtual void SetUser(IPrincipal principal)
{   
    if (!CurrentPrincipal.Equals(principal))
    {
        CurrentPrincipal = principal;
        if (CurrentPrincipal is ClaimsPrincipal)
        {
            SafeSetCurrentUser(AuthenticationStateProvider, (ClaimsPrincipal)CurrentPrincipal);
        }
        else
        {
            //??Throw some exception here??
        }
    }
}

```